controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -532,6 +532,126 @@ public void createScope(ScopeInfo request, StreamObserver<CreateScopeStatus> res
                 responseObserver, requestTag);
     }
 
+    @Override
+    public void listScopes(Controller.ScopesRequest request, StreamObserver<Controller.ScopesResponse> responseObserver) {
+        RequestTag requestTag = requestTracker.initializeAndTrackRequestTag(requestIdGenerator.get(),
+                "listScopes");
+        log.info(requestTag.getRequestId(), "listScope called.");
+
+        final AuthContext ctx;
+        if (this.grpcAuthHelper.isAuthEnabled()) {
+            ctx = AuthContext.current();
+        } else {
+            ctx = null;
+        }
+
+        Supplier<String> stringSupplier = () -> {
+            String result = this.grpcAuthHelper.checkAuthorization(
+                    AuthResourceRepresentation.ofScopes(),
+                    AuthHandler.Permissions.READ,
+                    ctx);
+            log.debug("Result of authorization for [{}] and READ permission is: [{}]",
+                    AuthResourceRepresentation.ofScopes(), result);
+            return result;
+        };
+        Function<String, CompletableFuture<Controller.ScopesResponse>> scopesFn = delegationToken -> controllerService
+                .listScopes(request.getContinuationToken().getToken(), pageLimit)
+                .thenApply(response -> {
+                    log.debug("All scopes in scope with continuation token: {}", response);
+                    return Controller.ScopesResponse
+                            .newBuilder().addAllScopes(response.getKey())
+                            .setContinuationToken(Controller.ContinuationToken.newBuilder()
+                                                                              .setToken(response.getValue()).build())
+                            .build();
+                });
+        authenticateExecuteAndProcessResults(stringSupplier, scopesFn, responseObserver, requestTag);
```

Comment:
Can the auth filter out some scopes, causing us to return less number of scopes than limit?

Reply:
We could, but i kept the authorization part here consistent with the equivalent api in the REST which requires users to have READ permission on "/" to be able to invoke list scopes. 

If we want a different approach for authorization, we should do it consistently on both interfaces. 
I have created a separate issue for the same here:
https://github.com/pravega/pravega/issues/4955